actionpack/lib/action_controller/metal/strong_parameters.rb

```diff
@@ -235,31 +248,99 @@ def ==(other)
     end
 
     # Returns a safe <tt>ActiveSupport::HashWithIndifferentAccess</tt>
-    # representation of this parameter with all unpermitted keys removed.
+    # representation of the parameters with all unpermitted keys removed.
     #
     #   params = ActionController::Parameters.new({
-    #     name: 'Senjougahara Hitagi',
-    #     oddity: 'Heavy stone crab'
+    #     name: "Senjougahara Hitagi",
+    #     oddity: "Heavy stone crab"
     #   })
-    #   params.to_h # => {}
+    #   params.to_h
+    #   # => ActionController::UnfilteredParameters: unable to convert unfiltered parameters to hash
     #
     #   safe_params = params.permit(:name)
     #   safe_params.to_h # => {"name"=>"Senjougahara Hitagi"}
     def to_h
       if permitted?
         convert_parameters_to_hashes(@parameters, :to_h)
+      elsif self.class.raise_on_unfiltered_parameters
+        raise UnfilteredParameters
       else
         slice(*self.class.always_permitted_parameters).permit!.to_h
       end
     end
 
+    # Returns a safe <tt>Hash</tt> representation of the parameters
+    # with all unpermitted keys removed.
+    #
+    #   params = ActionController::Parameters.new({
+    #     name: "Senjougahara Hitagi",
+    #     oddity: "Heavy stone crab"
+    #   })
+    #   params.to_hash
+    #   # => ActionController::UnfilteredParameters: unable to convert unfiltered parameters to hash
+    #
+    #   safe_params = params.permit(:name)
+    #   safe_params.to_hash # => {"name"=>"Senjougahara Hitagi"}
+    def to_hash
+      if self.class.raise_on_unfiltered_parameters
+        to_h.to_hash
+      else
+        unless permitted?
+          message = <<-DEPRECATE.squish
+          Method #to_hash when `config.action_controller.raise_on_unfiltered_parameters` is false
+          is deprecated and will be removed in Rails 5.1 when it is going to raise if the
+          `ActionController::Parameters` is unfiltered. Using this deprecated behavior exposes
+          potential security problems. If you continue to use this method you may be creating
+          a security vulnerability in your app that can be exploited. Instead,
+          consider enabling `config.action_controller.raise_on_unfiltered_parameters` and filtering
+          the `ActionController::Parameters`.
+          DEPRECATE
```

Comment:
> to_hash unexpectedly ignores parameter filtering, and will change to enforce it in Rails 5.1. Enable raise_on_unfiltered_parameters to respect parameter filtering, which is the default in new applications. For the existing deprecated behaviour, call to_unsafe_h instead.

Maybe? The "will be removed" phrasing seems particularly awkward when it's not really going away, per se.